### PR TITLE
Add warning when leaving form with unsaved changes

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] Display warning when leaving form with unsaved changes
- [x] Redisplay form with unsaved changes on cancel
- [x] Make it work when the user navigates away from the form using the back button
- [x] Add tests
- [x] Add documentation

Peek 23-03-2020 10-02

### CODE REVIEW

File: packages/ra-core/src/form/useWarnWhenUnsavedChanges.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    it('should not warn when leaving form with submit button', () => {
+        const { getByLabelText, getByText } = render(<App />);
+        const input = getByLabelText('First Name') as HTMLInputElement;
+        input.value = 'John Doe';
+        fireEvent.click(getByText('Submit'));
+        getByText('Submitted');
+    });
+
+    it('should warn when leaving form with no unsaved changes', () => {
```
Comment (djhi) [inline packages/ra-core/src/form/useWarnWhenUnsavedChanges.spec.tsx]:
```suggestion
    it('should warn when leaving form with unsaved changes', () => {
```

File: packages/ra-core/src/form/useWarnWhenUnsavedChanges.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            'ra.message.unsaved_changes'
+        );
+        // check that we're still in the form and that the unsaved changes are here
+        expect((getByLabelText('First Name') as HTMLInputElement).value).toBe(
+            'John Doe'
+        );
+        expect(queryByText('Somewhere')).toBeNull();
+    });
+
+    it('should warn when leaving form with no unsaved changes but accept override', () => {
```
Comment (djhi) [inline packages/ra-core/src/form/useWarnWhenUnsavedChanges.spec.tsx]:
```suggestion
    it('should warn when leaving form with unsaved changes but accept override', () => {
```

File: packages/ra-core/src/form/useWarnWhenUnsavedChanges.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Display a confirmation dialog if the form has unsaved changes.
+ * - If the user confirms, the navigation continues and the changes are lost.
+ * - If the user cancels, the navigation is reverted and the changes are kept.
+ *
+ * We can't use history.block() here because forms have routes, too (for
+ * instance TabbedForm), and the confirm dialog would show up when navigating
+ * inside the form. So instead of relying on route change detection, we rely
+ * on unmount detection. The resulting UI isn't perfect, because when they
+ * click the cancel buttun, users briefly see the page they asked before
```
Comment (djhi) [inline packages/ra-core/src/form/useWarnWhenUnsavedChanges.tsx]:
```suggestion
 * click the cancel button, users briefly see the page they asked before
```

File: packages/ra-language-french/src/index.ts [hunk 4]
```diff
@@ -90,6 +90,8 @@ const frenchMessages: TranslationMessages = {
             not_found:
                 "L'URL saisie est incorrecte, ou vous avez suivi un mauvais lien.",
             yes: 'Oui',
+            unsaved_changes:
+                "Certains changements n'ont pas été enregistrés. Etes-vous sûr(e) de vouloir quitter cette page ?",
```
Comment (djhi) [inline packages/ra-language-french/src/index.ts]:
```suggestion
                "Certains changements n'ont pas été enregistrés. Êtes-vous sûr·e de vouloir quitter cette page ?",
```

Comment (fzaninotto) [reply packages/ra-language-french/src/index.ts]: we have "sûr(e)" already in other translations, let's keep parentheses for this one.

**Pull request merged**